### PR TITLE
snapgene-viewer: Add package

### PR DESCRIPTION
**Pull request by Kiri2002**
## Involved packages

 - snapgene-viewer

## Involved issue
Add package

Close #

## Details

- [x] Tested in the local machine (largest 16G RAM) and it is passed without any issue
- [x] Provide New Package
- [ ] Fix the Packages
  - [ ] PKGBUILD
  - [ ] lilac.yaml
  - [ ] lilac.py
- [x] Would like to continue to work with us

#### snapgene-viewer
- [snapgene-viewer-7.2.0-1-x86_64-package.log](https://github.com/BioArchLinux/Packages/files/14822729/snapgene-viewer-7.2.0-1-x86_64-package.log)
- [snapgene-viewer-7.2.0-1-x86_64.pkg.tar.zst-namcap.log](https://github.com/BioArchLinux/Packages/files/14822730/snapgene-viewer-7.2.0-1-x86_64.pkg.tar.zst-namcap.log)

The ELF error caused by upstream  can't be fixed. This is a close source software like mega.

### CODE REVIEW

File: BioArchLinux/snapgene-viewer/PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+         'libglvnd'
+         'libxext'
+         'openssl-1.1'
+         'zlib'
+         'qca-qt6'
+         'freetype2'
+         'libxkbcommon'
+         'sh'
+)
+makedepends=('patchelf')
```
Comment (starsareintherose) [inline BioArchLinux/snapgene-viewer/PKGBUILD:52]: You don't have build functions, why you use makedepends?

**Review by starsareintherose — CHANGES_REQUESTED** [review]

File: BioArchLinux/snapgene-viewer/PKGBUILD (hunk 1 above)
Comment (Kiri2002) [reply BioArchLinux/snapgene-viewer/PKGBUILD:52]: Fix rpath using patchelf in the end of package()

**Comment by Kiri2002** [comment]
The original .so files and snapgene(bin) has wrong rpath forward to "/home/centos", example:
```
snapgene E: Insecure RPATH '/home/centos/.conan/data/qt-for-sg/6.2.10/snapgene/stable/package/f6a1919c5d86e9463ebddc2f3d4066e8e92eadce/lib' in file ('opt/gslbiotech/snapgene/libqntp.so.1.0.0')
snapgene E: Insecure RPATH '/home/centos/.conan/data/qt-for-sg/6.2.10/snapgene/stable/package/f6a1919c5d86e9463ebddc2f3d4066e8e92eadce/lib' in file ('opt/gslbiotech/snapgene/libqwebdavlib.so.1.0.0')
```
So I use "patchelf" to fix it.

File: BioArchLinux/snapgene-viewer/PKGBUILD (hunk 1 above)
Comment (starsareintherose) [reply BioArchLinux/snapgene-viewer/PKGBUILD:52]: how about creating a install file?

**Comment by starsareintherose** [comment]
@hubutui @lilydjwg @pekkarr would you like to give a review for this pr

**Comment by Kiri2002** [comment]
The "rpath" mean "run path", is parameter for .so or executable bin etc.
It will try to access to "rpath" to find libs when run.

> why it creates centos home dir?

It not create any dir. And why some file have wrong rpath, the upstream should be blamed who built it in wrong way.

> how about creating a install file?

I searched in Archlinuxcn, and don't find any example to fix rpah in install/patch file. But in PKGBUILD.

https://github.com/search?q=repo%3Aarchlinuxcn%2Frepo%20patchelf&type=code

File: BioArchLinux/snapgene-viewer/PKGBUILD [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# Maintainer: Bitals <me at bitals dot xyz>
+# Contributor: Matthijs Tadema <M dot J dot Tadema at gmail dot com>
+# Contributor: Lorenzo Gaifas <brisvag at gmail dot com>
+# Contributor: Antony Lee <anntzer dot lee at gmail dot com>
+pkgname=snapgene-viewer
+pkgver=7.2.0
+_pkgver_major=$(cut -d '.' -f 1 <<<"$pkgver")
```
Comment (lilydjwg) [inline BioArchLinux/snapgene-viewer/PKGBUILD:7]:
```suggestion
_pkgver_major=${pkgver%%.*}
```

**Comment by Kiri2002** [comment]
@starsareintherose  There are more information about "rpath":
https://en.wikipedia.org/wiki/Rpath
https://bugs.archlinux.org/task/14049

File: BioArchLinux/snapgene-viewer/PKGBUILD [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+_pkgver_major_middle=$(cut -d '.' -f 1-2 <<<"$pkgver")
```
Comment (lilydjwg) [inline BioArchLinux/snapgene-viewer/PKGBUILD:8]:
```suggestion
_pkgver_major_middle=${pkgver%.*}
_pkgver_major_middle=${_pkgver_major_middle#*.}
```

File: BioArchLinux/snapgene-viewer/PKGBUILD (hunk 1 above)
Comment (lilydjwg) [reply BioArchLinux/snapgene-viewer/PKGBUILD:52]: > how about creating a install file?

Not good. Install scripts are run on installation, not packaging.

File: BioArchLinux/snapgene-viewer/PKGBUILD [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+package() {
+    cd "$pkgdir"
+    cp -r "$srcdir/opt" "$pkgdir"
+    cp -r "$srcdir/usr" "$pkgdir"
+
+    install -Dm 755 "$srcdir/snapgene-viewer.sh" "$pkgdir/usr/bin/snapgene-viewer"
+    install -Dm 664 "$pkgdir/opt/gslbiotech/snapgene-viewer/resources/licenseAgreement.html" "$pkgdir/usr/share/licenses/$pkgname/LICENSE.html"
+
+    sed -i 's`${INSTALLED_DIR}/snapgene-viewer "$@"`QT_QPA_PLATFORM="xcb" ${INSTALLED_DIR}/snapgene-viewer "$@"`' "$pkgdir/opt/gslbiotech/snapgene-viewer/snapgene-viewer.sh"
```
Comment (lilydjwg) [inline BioArchLinux/snapgene-viewer/PKGBUILD:67]:
```suggestion
    sed -i 's|${INSTALLED_DIR}/snapgene-viewer "$@"|QT_QPA_PLATFORM="xcb" ${INSTALLED_DIR}/snapgene-viewer "$@"|' "$pkgdir/opt/gslbiotech/snapgene-viewer/snapgene-viewer.sh"
```

` is too confusing and uncommon for this purpose.

File: BioArchLinux/snapgene/PKGBUILD [hunk 5]
```diff
@@ -0,0 +1,93 @@
+# Based on snapgene-viewer package by Antony Lee <anntzer dot lee at gmail dot com>
+# Maintainer: Bitals <me at bitals dot xyz>
+# Contributor: Matthijs Tadema <M dot J dot Tadema at gmail dot com>
+# Contributor: Lorenzo Gaifas <brisvag at gmail dot com>
+pkgname=snapgene
+pkgver=7.2.0
+_pkgver_major=$(cut -d '.' -f 1 <<<"$pkgver")
```
Comment (lilydjwg) [inline BioArchLinux/snapgene/PKGBUILD:7]: ditto

**Comment by starsareintherose** [comment]
If @lilydjwg agree, we are delighted to merge it

**Comment by hubutui** [comment]
Is this pkg free software? Could we redistribute it?

**Comment by starsareintherose** [comment]
It seems that this pkg is commercial one.

**Comment by Kiri2002** [comment]
Just like mega

**Review by lilydjwg — APPROVED** [review]

**Comment by Kiri2002** [comment]
image

Both them can download without money.
But snapgene need license(can try for free for some days).

**Comment by Kiri2002** [comment]
image

Should the ms-font be added to opt-depends?

File: BioArchLinux/snapgene-viewer/PKGBUILD [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+         'zlib'
+         'qca-qt6'
+         'freetype2'
+         'libxkbcommon'
+         'sh'
+)
+makedepends=('patchelf')
+# A valid licence is required to use the full version of snapgene
+source=("https://cdn.snapgene.com/downloads/SnapGeneViewer/"$_pkgver_major".x/"$_pkgver_major_middle"/"$pkgver"/snapgene_viewer_"$pkgver"_linux.rpm"
+        "snapgene-viewer.sh")
```
Comment (hubutui) [inline BioArchLinux/snapgene-viewer/PKGBUILD:56]: source file `snapgene-viewer.sh` is missing.

File: BioArchLinux/snapgene/PKGBUILD [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+         'zlib'
+         'qca-qt6'
+         'freetype2'
+         'libxkbcommon'
+         'sh')
+
+makedepends=('patchelf')
+# A valid licence is required to use the full version of snapgene
+source=("https://cdn.snapgene.com/downloads/SnapGene/"$_pkgver_major".x/"$_pkgver_major_middle"/"$pkgver"/"$pkgname"_"$pkgver"_linux.rpm"
+        "snapgene.sh")
```
Comment (hubutui) [inline BioArchLinux/snapgene/PKGBUILD:56]: where is `snapgene.sh`?

**Comment by Kiri2002** [comment]
sorry for forget upload *.sh. Added now

**Comment by hubutui** [comment]
could you check the source again?
```text
 updpkgsums
==> Retrieving sources...
  -> Downloading snapgene_7.2.0_linux.rpm...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (56) The requested URL returned error: 404
==> ERROR: Failure while downloading https://cdn.snapgene.com/downloads/SnapGene/7.x/2/7.2.0/snapgene_7.2.0_linux.rpm
    Aborting...
==> ERROR: Failed to generate new checksums

==> Retrieving sources...
  -> Downloading snapgene_viewer_7.2.0_linux.rpm...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (56) The requested URL returned error: 404
==> ERROR: Failure while downloading https://cdn.snapgene.com/downloads/SnapGeneViewer/7.x/2/7.2.0/snapgene_viewer_7.2.0_linux.rpm
    Aborting...
==> ERROR: Failed to generate new checksums
```

**Comment by hubutui** [comment]
you can follow this [wiki](https://wiki.archlinux.org/title/DeveloperWiki:Building_in_a_clean_chroot) to build this pkg in a clean chroot.

**Comment by Kiri2002** [comment]
> could you check the source again?
> 
> ```
>  updpkgsums
> ==> Retrieving sources...
>   -> Downloading snapgene_7.2.0_linux.rpm...
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (56) The requested URL returned error: 404
> ==> ERROR: Failure while downloading https://cdn.snapgene.com/downloads/SnapGene/7.x/2/7.2.0/snapgene_7.2.0_linux.rpm
>     Aborting...
> ==> ERROR: Failed to generate new checksums
> 
> ==> Retrieving sources...
>   -> Downloading snapgene_viewer_7.2.0_linux.rpm...
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (56) The requested URL returned error: 404
> ==> ERROR: Failure while downloading https://cdn.snapgene.com/downloads/SnapGeneViewer/7.x/2/7.2.0/snapgene_viewer_7.2.0_linux.rpm
>     Aborting...
> ==> ERROR: Failed to generate new checksums
> ```

this issue cause by this:
image
now fixed.

**Comment by starsareintherose** [comment]
> image
> 
> Should the ms-font be added to opt-depends? It use system font now, and work fine

It doesn't make sense, some commercial software is free to use, but they don't want us to re-distribute it

**Comment by Kiri2002** [comment]
> > image
> > Should the ms-font be added to opt-depends? It use system font now, and work fine
> 
> It doesn't make sense, some commercial software is free to use, but they don't want us to re-distribute it

Maybe just do it, remove when the vendor ask us to do.

**Comment by starsareintherose** [comment]
@hubutui final check, then you can squash and merge

**Comment by starsareintherose** [comment]
Final decision: only snapegene-viewer can be merged

**Comment by Kiri2002** [comment]
> Final decision: only snapegene-viewer can be merged

snapgene separated to https://github.com/BioArchLinux/Packages/pull/204, for maybe something changed in the future.

**Review by starsareintherose — APPROVED** [review]

**Pull request merged**